### PR TITLE
Update to Soldeer 0.3.0

### PR DESCRIPTION
**Pull request by mario-eth**
## Motivation
Soldeer 0.3.0 will be released which has a chunk of it rewritten to be more idiomatic.

## Solution
Big refactor to adopt idiomatic Rust patterns and improve the existing public API.

### Config file
The config file (whichever has a `[dependencies]` table between `foundry.toml` and `soldeer.toml`) now has a `[soldeer]` section with the following format and defaults:

```toml
[soldeer]
# whether soldeer manages remappings
remappings_generated = true

# whether soldeer re-generates all remappings when installing, updating or uninstalling deps
remappings_regenerate = false

# whether to suffix the remapping with the version: `name-a.b.c`
remappings_version = true

# a prefix to add to the remappings ("@" would give `@name`)
remappings_prefix = ""

# where to store the remappings ("txt" for `remappings.txt` or "config" for `foundry.toml`)
# ignored when `soldeer.toml` is used as config (uses `remappings.txt`)
remappings_location = "txt"
```

### Remappings

Remappings are now properly managed by Soldeer if desired. They can be updated inside the `foundry.toml` file if that file is also used for `[dependencies]`.

Remappings in the `[profile.default]` are always managed if `remappings_generated = true`. Other profiles only get managed if they already contain a `remappings` array.

If `remappings_regenerate = false`, then existing remappings are not modified by Soldeer, unless a dependency is removed. This allows for remappings customization.

### Download sub-dependencies
Dependencies of dependencies are now automatically pulled into the main dependencies. For example, if `dep1` has submodules or Soldeer dependencies, using the `--recursive-deps` option or setting `recursive-deps = true` under `[soldeer]` in the TOML config file will automatically download the sub-dependencies.

More details in the Soldeer PRs
https://github.com/mario-eth/soldeer/pull/118
https://github.com/mario-eth/soldeer/pull/136

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, could we group the soldeer remapping settings into a table so that this can be configured via `[soldeer.remappings]`?

**Comment by mario-eth** [comment]
> cool, could we group the soldeer remapping settings into a table so that this can be configured via `[soldeer.remappings]`?

But if we do this, how does foundry manages it for the default profile?

**Comment by mattsse** [comment]
hmm, these belong to the soldeer table? so we can nest them in another table just for remapping settings?

**Review by ArkFunds — APPROVED** [review]

**Comment by mario-eth** [comment]
> hmm, these belong to the soldeer table? so we can nest them in another table just for remapping settings?

hmm maybe i missunderstood you mean to have like 
```TOML
[soldeer.remappings]
generated = true
regenerate = true
location = "config"
prefix = "@"
``` 

The problem is that we have also `recursive-deps` that it's part of the `soldeer`

so then the full toml will be 
```TOML
[soldeer.remappings]
generated = true
regenerate = true
location = "config"
prefix = "@"
[soldeer]
recursive-deps = true
```

File: crates/config/src/soldeer.rs
Comment (mattsse) [inline crates/config/src/soldeer.rs:1]: could we add some sanity Config parsing tests that visualize this, similar to tests in config/lib.rs

File: crates/forge/bin/cmd/soldeer.rs
Comment (mattsse) [inline crates/forge/bin/cmd/soldeer.rs:1]: could we also add some docs here about how the soldeer commands makes use of the config settings, this step is a bit opaque to me rn

**Review by mattsse — CHANGES_REQUESTED** [review]
left some requests for docs.  I don't really mind if we put all remapping settings directly into the soldeer object

**Comment by 0xCalibur** [comment]
Can't wait to use it! Does this support not-having-to prepend the version on dependencies folder name.
for example if I install solady 0.1, I just prefer having the folder named solady and not solady-0.1. This way I don't need to change my remappings config...

**Comment by beeb** [comment]
@0xCalibur yes this is supported in this version!

**Comment by mario-eth** [comment]
@0xCalibur Important to know is that if you set  `remappings_version = true` then your folder name won't actually change but your remappings entry will not have the version. This will be the exact behaviour that you want. 
e.g. for solady in your case the remappings entry will look like:
`solady/=dependencies/solady-0.1/`

You won't need to change anything anymore even if you update your solady version let's say to 0.2 it will automatically do
`solady/=dependencies/solady-0.2/` which will result in you not needing to change this in the files.

File: crates/config/src/soldeer.rs
Comment (mario-eth) [reply crates/config/src/soldeer.rs:1]: Check them out now

File: crates/forge/bin/cmd/soldeer.rs
Comment (mario-eth) [reply crates/forge/bin/cmd/soldeer.rs:1]: Added docs and made the configs more explicit

**Review by mattsse — APPROVED** [review]
pending CI

**Comment by 0xCalibur** [comment]
> @0xCalibur Important to know is that if you set `remappings_version = true` then your folder name won't actually change but your remappings entry will not have the version. This will be the exact behaviour that you want. e.g. for solady in your case the remappings entry will look like: `solady/=dependencies/solady-0.1/`
> 
> You won't need to change anything anymore even if you update your solady version let's say to 0.2 it will automatically do `solady/=dependencies/solady-0.2/` which will result in you not needing to change this in the files.

Thanks for answering.

The thing is that I rather manage the remappings myself. See https://github.com/Abracadabra-money/abracadabra-money-contracts/blob/main/foundry.toml

My preferred behavior would be that it just install the dependencies in the same folder name over the versions and my remapping config would stay the same.

I believe the soldeer config file is clear enough about the current used version.

**Comment by 0xCalibur** [comment]
Btw, I'm using my custom "package management" for now, looking to switch over soldeer.

I'm also doing this to check for lib changes.
See: https://github.com/Abracadabra-money/abracadabra-money-contracts/blob/main/tooling/tasks/core/check-libs-integrity.ts

This way, if for some reason, the lib files have been changed/compromised/altered by doing debugging with them, or for example, you open them and change a file by mistake, it's won't build/deploy.

Also since dependencies are git ignored, it's really a safety net. Of course, tests should be ran before deploying but I think such check is neat. wdyt?

**Comment by beeb** [comment]
> I'm also doing this to detect for lib changes: See: https://github.com/Abracadabra-money/abracadabra-money-contracts/blob/main/tooling/tasks/core/check-libs-integrity.ts

Hey @0xCalibur, this has already been implemented by me for 0.4.0 :)

**Comment by mario-eth** [comment]
> > @0xCalibur Important to know is that if you set `remappings_version = true` then your folder name won't actually change but your remappings entry will not have the version. This will be the exact behaviour that you want. e.g. for solady in your case the remappings entry will look like: `solady/=dependencies/solady-0.1/`
> > You won't need to change anything anymore even if you update your solady version let's say to 0.2 it will automatically do `solady/=dependencies/solady-0.2/` which will result in you not needing to change this in the files.
> 
> Thanks for answering.
> 
> The thing is that I rather manage the remappings myself. See https://github.com/Abracadabra-money/abracadabra-money-contracts/blob/main/foundry.toml
> 
> My preferred behavior would be that it just install the dependencies in the same folder name over the versions and my remapping config would stay the same.
> 
> I believe the soldeer config file is clear enough about the current version used.

Got it, for now we decided to just keep the left side of the remappings unchanged via a config option, we will think about the dependency directory as well if we will enable it via a certain flag in the next version. We are playing with supporting multiple versions for the same dependency which will require different directories.

**Comment by mario-eth** [comment]
as well @0xCalibur feel free to open an issue on: https://github.com/mario-eth/soldeer

**Review by mattsse — APPROVED** [review]

**Pull request merged**